source/h2_frames.c

```diff
@@ -0,0 +1,1618 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/http/private/h2_frames.h>
+#include <aws/http/private/hpack.h>
+
+#include <aws/http/request_response.h>
+
+#include <aws/compression/huffman.h>
+
+/* Stream ids & dependencies should only write the bottom 31 bits */
+static const uint32_t s_31_bit_mask = UINT32_MAX >> 1;
+static const uint32_t s_u32_top_bit_mask = UINT32_MAX << 31;
+
+static const uint8_t s_indexed_header_field_mask = 1 << 7;
+static const uint8_t s_literal_save_field_mask = 1 << 6;
+static const uint8_t s_dynamic_table_size_update_mask = 1 << 5;
+static const uint8_t s_literal_no_forward_save_mask = 1 << 4;
+
+/* RFC-7540 6.5.2 */
+static const size_t s_hpack_dynamic_table_initial_size = 4096;
+/* TBD */
+static const size_t s_hpack_dynamic_table_max_size = 4096;
+
+/***********************************************************************************************************************
+ * Priority
+ **********************************************************************************************************************/
+static size_t s_frame_priority_settings_size = 5;
+
+static int s_frame_priority_settings_encode(
+    const struct aws_h2_frame_priority_settings *priority,
+    struct aws_byte_buf *output) {
+    AWS_PRECONDITION(priority);
+    AWS_PRECONDITION(output);
+    AWS_FATAL_ASSERT((priority->stream_dependency & s_u32_top_bit_mask) == 0);
+
+    /* Write the top 4 bytes */
+    uint32_t top_bytes = priority->stream_dependency | ((uint32_t)priority->stream_dependency_exclusive << 31);
+    if (!aws_byte_buf_write_be32(output, top_bytes)) {
+        return AWS_OP_ERR;
+    }
+
+    /* Write the priority weight */
+    if (!aws_byte_buf_write_u8(output, priority->weight)) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_frame_priority_settings_decode(
+    struct aws_h2_frame_priority_settings *priority,
+    struct aws_h2_frame_decoder *decoder) {
+    AWS_PRECONDITION(priority);
+    AWS_PRECONDITION(decoder);
+
+    /* Read the top 4 bytes */
+    uint32_t top_bytes = 0;
+    if (!aws_byte_cursor_read_be32(&decoder->payload, &top_bytes)) {
+        return AWS_OP_ERR;
```

Comment:
likewise, byte_cursor_read doesn't raise an error
this applies to maaaany places